### PR TITLE
hashes: Import of SHA-1 algorithm

### PR DESCRIPTION
**Pull request by OlegHahm**
I compared five different implementations:
* [OAUTH](http://oauth.googlecode.com/svn/code/c/liboauth/src/sha1.c) - public domain
* [ORYX](http://www.oryx-embedded.com/doc/sha1_8h.html#ab734cb0e0b67798f5cce1f34419982a3) - commercial, GPLv2
* [PD](http://www.ghostscript.com/doc/jbig2dec/sha1.c) - public domain
* [RFC](https://tools.ietf.org/html/rfc3174#section-7) - [IETF rights in contribution](http://tools.ietf.org/html/rfc3978)
* [LIBSHA1](https://github.com/dottedmag/libsha1) - BSD license

The size results for my [example application](https://github.com/OlegHahm/miniature-dangerzone/tree/master/sha1-test) revealed that the OAUTH implementation has the smallest memory footprint of all compared implementations for all tested boards:
```
   text     bss    data   dec      hex

OAUTH
   9072     184    2760   12016    2ef0 samr21-xpro/sha1.elf
ORYX
   9168     184    2760   12112    2f50 samr21-xpro/sha1.elf
PD
  12944     184    2760   15888    3e10 samr21-xpro/sha1.elf
RFC
   9344     184    2760   12288    3000 samr21-xpro/sha1.elf
LIBSHA1
  12896     184    2760   15840    3de0 samr21-xpro/sha1.elf


OAUTH
   9000     184    2752   11936    2ea0 iotlab-m3/sha1.elf
ORYX
   9072     184    2752   12008    2ee8 iotlab-m3/sha1.elf
PD
  12496     184    2752   15432    3c48 iotlab-m3/sha1.elf
RFC
   9240     184    2752   12176    2f90 iotlab-m3/sha1.elf
LIBSHA1
  12364     184    2752   15300    3bc4 iotlab-m3/sha1.elf

OAUTH
   7936      68    1058    9062    2366 z1/sha1.elf
ORYX
   8054      68    1058    9180    23dc z1/sha1.elf
PD
  25806      68    1058   26932    6934 z1/sha1.elf
RFC
   8152      68    1058    9278    243e z1/sha1.elf
LIBSHA1
  24810      68    1058   25936    6550 z1/sha1.elf

OAUTH
   8896     420     720   10036    2734 arduino-mega2560/sha1.elf
ORYX
   8912     484     720   10116    2784 arduino-mega2560/sha1.elf
PD
  28692     424     720   29836    748c arduino-mega2560/sha1.elf
RFC
   9906     420     720   11046    2b26 arduino-mega2560/sha1.elf
LIBSHA1
  29040     420     720   30180    75e4 arduino-mega2560/sha1.elf
```

### CODE REVIEW

**Comment by OlegHahm** [comment]
The sha1 implementation alone requires about 540 bytes of ROM on Cortex platforms and about 1,2kB on AVR8 and MSP430 platforms.

**Comment by OlegHahm** [comment]
Compared to our MD5 implementation that's about 150 bytes less on AVR8 and MSP430 and about 440 bytes less on Cortex.

**Comment by BytesGalore** [comment]
and against `crypto/sha256`?

**Comment by OlegHahm** [comment]
Against SHA256 it requires about 1.1kB less on AVR8, 550 bytes less on MSP430 and 32 bytes less on Cortex-M0.

**Comment by authmillenon** [comment]
How big is the sha1.o?

**Comment by OlegHahm** [comment]
https://github.com/RIOT-OS/RIOT/pull/4701#issuecomment-175656626

File: sys/hashes/sha1.c [hunk 1]
```diff
@@ -0,0 +1,134 @@
+/* This code is public-domain - it is based on libcrypt
```
Comment (haukepetersen) [inline sys/hashes/sha1.c]: Could you also add a doxygen header into this file? It might be of interest later who imported it and so on... Thx

**Comment by haukepetersen** [comment]
Looking good! ACK once the missing header is added.

**Comment by OlegHahm** [comment]
added the desired information to C file. Okay to squash?

**Comment by OlegHahm** [comment]
added a convenient function similar to the one that the other hashing functions offer. Ready to squash?

**Comment by BytesGalore** [comment]
Don't know how to PR against your branch so maybe you could add hmac too [1]
[1] https://github.com/BytesGalore/RIOT/tree/add_hmac_sha1

File: sys/hashes/sha1.c [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    /* Return pointer to hash (20 characters) */
+    return (uint8_t *) s->state;
+}
+
+void sha1(uint8_t *dst, const uint8_t *src, size_t len)
+{
+    sha1_context ctx;
+    sha1_init(&ctx);
+    sha1_update(&ctx, (unsigned char*) src, len);
+    memcpy(dst, sha1_final(&ctx), SHA1_DIGEST_LENGTH);
```
Comment (BytesGalore) [inline sys/hashes/sha1.c]: This works since the digest is pushed on the stack, but I'm a bit nervous looking at this

**Comment by BytesGalore** [comment]
consider hmac for sha1 https://github.com/OlegHahm/RIOT/pull/14

File: sys/hashes/sha1.c (hunk 2 above)
Comment (OlegHahm) [inline sys/hashes/sha1.c]: Would you prefer `memcpy(dst, ctx->state, SHA1_DIGEST_LENGTH)`?

Comment (BytesGalore) [inline sys/hashes/sha1.c]: just personal taste, yes.
But no objection if you keep it the way it is.

**Comment by OlegHahm** [comment]
I added the hmac part from the original oauth code plus test vectors from [RFC2202](https://tools.ietf.org/html/rfc2202).

**Comment by Kijewski** [comment]
Some other current PR adds HMAC SHA-256. The algorithm should be generalizable to accept any hashing function. If it does not add to much overhead to have a generic function, then I would opt for it instead of basically adding the same algo twice.

**Comment by BytesGalore** [comment]
+1 to generalize `hmac`, but then we should move the `crypto/sha256` implementation here and also unify the function definitions.
Since I'm the one requested you to add `hmac` in `sha1`, I can try to make a generic `hashes/hmac` and move `crypto/sha256` here.
Sorry for all the OT.

Beside that your implementation looks good and works as expected.

**Comment by OlegHahm** [comment]
Before we generalize hmac, I would strongly vote for merging this PR first - with or without hmac.

**Comment by OlegHahm** [comment]
btw. we may rethink if we want to distinguish (directory-wise) between "simple" hashing functions and those which are considered to be "secure".

**Comment by OlegHahm** [comment]
And another btw: both sha-hmac1 implementations are comparable memory-wise while @BytesGalore's should consume less stack when sha1 without hmac is used.

**Comment by OlegHahm** [comment]
I'm not in favor of keeping this PR for much longer. Either merge with or without hmac or close it.

File: sys/hashes/sha1.c [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+    sha1_init(&ctx);
+    sha1_update(&ctx, (unsigned char *) src, len);
+    memcpy(dst, sha1_final(&ctx), SHA1_DIGEST_LENGTH);
+}
+
+#define HMAC_IPAD 0x36
+#define HMAC_OPAD 0x5c
+
+void sha1_init_hmac(sha1_context *s, const unsigned char *key, int keyLength)
```
Comment (Kijewski) [inline sys/hashes/sha1.c]: `keyLength` should be `unsigned`, probably `size_t`.

File: sys/hashes/sha1.c [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+void sha1_update(sha1_context *s, const unsigned char *data, size_t len)
+{
+    while (len--) {
+        sha1_update_byte(s, *(data++));
+    }
+}
+
+void sha1_pad(sha1_context *s)
```
Comment (Kijewski) [inline sys/hashes/sha1.c]: `static`

File: sys/hashes/sha1.c [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    b[s->buffer_offset ^ 3] = data;
+#endif
+    s->buffer_offset++;
+    if (s->buffer_offset == SHA1_BLOCK_LENGTH) {
+        sha1_hash_block(s);
+        s->buffer_offset = 0;
+    }
+}
+
+void sha1_update_byte(sha1_context *s, unsigned char data)
```
Comment (Kijewski) [inline sys/hashes/sha1.c]: `static`

File: sys/hashes/sha1.c [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        a = t;
+    }
+    s->state[0] += a;
+    s->state[1] += b;
+    s->state[2] += c;
+    s->state[3] += d;
+    s->state[4] += e;
+}
+
+void sha1_add_uncounted(sha1_context *s, uint8_t data)
```
Comment (Kijewski) [inline sys/hashes/sha1.c]: `static`

File: sys/hashes/sha1.c [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    s->byte_count = 0;
+    s->buffer_offset = 0;
+}
+
+uint32_t sha1_rol32(uint32_t number, uint8_t bits)
+{
+    return ((number << bits) | (number >> (32 - bits)));
+}
+
+void sha1_hash_block(sha1_context *s)
```
Comment (Kijewski) [inline sys/hashes/sha1.c]: `static`

File: sys/hashes/sha1.c [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    s->state[0] = 0x67452301;
+    s->state[1] = 0xefcdab89;
+    s->state[2] = 0x98badcfe;
+    s->state[3] = 0x10325476;
+    s->state[4] = 0xc3d2e1f0;
+    s->byte_count = 0;
+    s->buffer_offset = 0;
+}
+
+uint32_t sha1_rol32(uint32_t number, uint8_t bits)
```
Comment (Kijewski) [inline sys/hashes/sha1.c]: `static`

**Comment by Kijewski** [comment]
But for my minor comments the code looks very fine. I did not find places to optimize or obvious attack vectors, but IANAC*. Why does Travis not pick up this PR?

<sup>* I am not a cryptographer.</sup>

**Comment by OlegHahm** [comment]
Thanks for the review, addressed your comments. Travis ignores this PR for now, because I put a `[ci skip]` into the commit messages.

**Comment by OlegHahm** [comment]
Ok to squash?

File: tests/unittests/tests-hashes/tests-hashes.c [hunk 9]
```diff
@@ -23,5 +23,5 @@
 void tests_hashes(void)
 {
     TESTS_RUN(tests_hashes_md5_tests());
-    TESTS_RUN(tests_hashes_sha256_tests());
+    TESTS_RUN(tests_hashes_sha1_tests());
```
Comment (Kijewski) [inline tests/unittests/tests-hashes/tests-hashes.c]: Bad rebase? Shouldn't this be an added line, not a replacement?

Comment (OlegHahm) [inline tests/unittests/tests-hashes/tests-hashes.c]: Oops, yes, thanks for the attentiveness.

**Comment by OlegHahm** [comment]
Fixed rebase errors.

**Comment by gebart** [comment]
@Kijewski @BytesGalore would you like to take over the assignment of this PR?

**Comment by BytesGalore** [comment]
@gebart ok, I do.

**Comment by BytesGalore** [comment]
While I don't like the undocumented magic like [1] which can be puzzled out digging into the standard [2], the code looks good (beside the magic voodoo) and works as expected, I would say please squash  and lets see what travis says.

[1] https://github.com/RIOT-OS/RIOT/pull/4701/files#diff-cf28151045a2c90554957d839d978296R65
[2] http://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf

**Comment by OlegHahm** [comment]
squashed

File: sys/include/hashes/sha1.h [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+void sha1(uint8_t *dst, const uint8_t *src, size_t len);
+
+/**
+ * @brief Initialize SHA-1 message digest context with MAC
+ *
+ * @param[in] s             Pointer to the SHA-1 context to initialize
+ * @param[in] key           Key used in the HMAC-SHA1 computation
+ * @param[in] key_length    The size in bytes of @p key
+ */
+void sha1_init_hmac(sha1_context *s, const unsigned char *key, int key_length);
```
Comment (BytesGalore) [inline sys/include/hashes/sha1.h]: nice, I wonder why my gcc didnt complained, please change `int key_length` to `size_t key_length`

Comment (OlegHahm) [inline sys/include/hashes/sha1.h]: fixed

**Comment by BytesGalore** [comment]
oh, you can directly squash

**Comment by OlegHahm** [comment]
did

File: sys/include/hashes/sha1.h (hunk 10 above)
Comment (BytesGalore) [inline sys/include/hashes/sha1.h]: ok, I overloaded the warning level of gcc to be basically turned off.

**Comment by OlegHahm** [comment]
@BytesGalore, do you ACK?

**Comment by BytesGalore** [comment]
Of course, I falsely assumed I already ACKed.
Still travis seems to be a unhappy with something, I kick him again.

**Comment by OlegHahm** [comment]
Well, you implicitly did, just wanted to be on the safe side.

Travis failed already yesterday, but I don't know what
```
$ ./dist/tools/travis-scripts/build_and_test.sh

First, rewinding head to replay your work on top of it...
Applying: hashes: add SHA-1 hashing algorithm
Applying: tests: added unittests for SHA-1
sys/include/hashes/sha1.h

The command "./dist/tools/travis-scripts/build_and_test.sh" exited with 2.
```
is supposed to tell me.

**Comment by BytesGalore** [comment]
oh, yup travis is very informative on this one.

**Comment by OlegHahm** [comment]
After running it three or four times, finally Travis revealed that it was unhappy with the documentation of the (internal) sha1 struct. Added documentation there.

File: sys/include/hashes/sha1.h [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+ **/
+typedef struct {
+    /** internal buffer */
+    uint32_t buffer[SHA1_BLOCK_LENGTH / sizeof(uint32_t)];
+    /** buffering current state of hashing */
+    uint32_t state[SHA1_DIGEST_LENGTH / sizeof(uint32_t)];
+    /** already processed bytes */
+    uint32_t byte_count;
+    /** I have no idea what this variable is good for */
+    uint8_t buffer_offset;
```
Comment (BytesGalore) [inline sys/include/hashes/sha1.h]: Internal variable used track if `buffer` is filled before proceed taking the hash this block [1].

[1] https://github.com/RIOT-OS/RIOT/pull/4701/files#diff-cf28151045a2c90554957d839d978296R105

Comment (OlegHahm) [inline sys/include/hashes/sha1.h]: Thanks

**Comment by OlegHahm** [comment]
Are you okay with my rephrasing of your proposed documentation line?

**Comment by BytesGalore** [comment]
yapp :+1:, please squash.
My ACK holds :)

**Comment by OlegHahm** [comment]
Thanks. Squashed. Let's see if Travis is happier now.

**Comment by BytesGalore** [comment]
Strange that T. randomly suppresses printing information why a build failed.

**Comment by OlegHahm** [comment]
Yes, I'm a bit concerned, too.

**Comment by OlegHahm** [comment]
But at least it's fine now. Thanks for the review!

**Pull request merged**